docs/src/tutorials/nonlinear/operator_ad.jl

```diff
@@ -324,3 +325,105 @@ function enzyme_rosenbrock()
 end
 
 enzyme_rosenbrock()
+
+# ## DifferentiationInterface
+
+# Julia offers [many different autodiff packages](https://juliadiff.org/).
+# [DifferentiationInterface.jl](https://github.com/gdalle/DifferentiationInterface.jl)
+# is a package that provides an abstraction layer across multiple underlying
+# autodiff libaries.
+
```

Comment:
Perhaps add a disclaimer:

While DI makes it easier to experiment with different autodiff tools, this comes at a cost. In particular, DI may produce slower derivatives than using an AD tool directly, and sometimes fail to work at all. For example, in this code passing AutoEnzyme to DI fails to differentiate the code which is differentiable with Enzyme directly like above.

Reply:
By the way do you have any idea why it fails here? As usual it's a segfault so not particularly user friendly

Reply:
No idea, I haven’t even tried to run it or seen what the error message, just took your word that it failed xD.

open an issue with the mwe and error and can take a look

Reply:
> and sometimes fail to work at all

I don't want to mention something in our documentation if this is the case. 

What is the failure mode? Will we gracefully report errors? Segfaults are not good.

Reply:
Oh agreed, in that case @gdalle just drop DI docs then?

Reply:
> open an issue with the mwe and error and can take a look

The MWE is literally this docs page but sure, I'll open an issue tomorrow on the Enzyme repo.

> I don't want to mention something in our documentation if this is the case.

To be fair, Enzyme will also sometimes fail to differentiate code, and throw very unpleasant errors. 

> What is the failure mode? Will we gracefully report errors? Segfaults are not good.

The fact that it is a segfault is not due to DI per se, it's Enzyme not catching the error gracefully. Sure, DI might be using Enzyme wrong in this particular case (perhaps because of a function returning two closures), but I too would very much prefer a clean error message from Enzyme to a segfault. Will try to reproduce tomorrow.

Reply:
> Oh agreed in that segfaults are bad, in that case @gdalle just drop DI docs then?

I'll try to debug the MWE with Enzyme, but it is by far the pickiest of backends, so I'm reasonably sure that it would work effortlessly if you plugged in Zygote or FiniteDiff. So I think there is still value in showcasing DI, with the proper warnings of course. @odow your call.

Reply:
I realized my changes mean we never actually mention Enzyme in the context of DI, so perhaps that's okay.

Reply:
> The fact that it is a segfault is not due to DI per se, it's Enzyme not catching the error gracefully. Sure, DI might be using Enzyme wrong in this particular case (perhaps because of a function returning two closures), but I too would very much prefer a clean error message from Enzyme to a segfault. Will try to reproduce tomorrow.

Again haven't seen the error message so not clear what's up, but I'm unaware of any segfaults caused by Enzyme atm. Most the time people saw segfaults in the past were actually bugs in the Julia compiler itself (e.g. https://github.com/JuliaLang/julia/pull/55306 ). 

That said unfortunately even if you're not the underlying cause of an error, the top level package you use is going to be the one people blame for issues (especially if code works otherwise without it). Analogously, there's a ton of code in Enzyme working around bugs in the Julia and LLVM compiler errors =/. Perhaps something similar would be good in DI?

Reply:
It is not DI's job to work around LLVM errors, it's Enzyme's. DI is a fancy argument-passer

Reply:
And here's the MWE: https://github.com/EnzymeAD/Enzyme.jl/issues/1942

Reply:
For the record, pure Enzyme also fails on this one when one uses `splat(f)`